src/main/scala/cognite/spark/v1/FileContentRelation.scala

```diff
@@ -0,0 +1,141 @@
+package cognite.spark.v1
+
+import cats.effect.std.Dispatcher
+import cats.effect.{IO, Resource}
+import com.cognite.sdk.scala.v1.{FileDownloadExternalId, GenericClient}
+import fs2.Stream
+import org.apache.commons.io.FileUtils
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.sources.{BaseRelation, Filter, PrunedFilteredScan, TableScan}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Row, SQLContext, SparkSession}
+import org.apache.spark.{Partition, TaskContext}
+import sttp.capabilities.WebSockets
+import sttp.capabilities.fs2.Fs2Streams
+import sttp.client3.asynchttpclient.SttpClientBackendFactory
+import sttp.client3.asynchttpclient.fs2.AsyncHttpClientFs2Backend
+import sttp.client3.{SttpBackend, UriContext, asStreamUnsafe, basicRequest}
+import sttp.model.Header
+
+trait WithSizeLimit {
+  val sizeLimit: Long
+}
+
+class FileContentRelation(config: RelationConfig, fileExternalId: String, inferSchema: Boolean)(
+    override val sqlContext: SQLContext)
+    extends BaseRelation
+    with TableScan
+    with PrunedFilteredScan
+    with Serializable
+    with WithSizeLimit {
+
+  override val sizeLimit: Long = 5 * FileUtils.ONE_GB
+
+  @transient lazy val client: GenericClient[IO] =
+    CdpConnector.clientFromConfig(config)
+
+  @transient private lazy val sttpFileContentStreamingBackendResource
+    : Resource[IO, SttpBackend[IO, Fs2Streams[IO] with WebSockets]] =
+    for {
+      dispatcher <- Dispatcher.parallel[IO]
+      backend <- Resource.make(
+        IO(
+          AsyncHttpClientFs2Backend
+            .usingClient[IO](SttpClientBackendFactory.create("file content download"), dispatcher))
+      )(backend => backend.close())
+    } yield backend
+
+  private lazy val dataFrame: DataFrame = createDataFrame(sparkSession = sqlContext.sparkSession)
+
+  override def schema: StructType =
+    dataFrame.schema
+
+  def createDataFrame(sparkSession: SparkSession): DataFrame = {
+    val rdd: RDD[String] = new RDD[String](sparkSession.sparkContext, Nil) with Serializable {
+
+      import cognite.spark.v1.CdpConnector.ioRuntime
+
+      val maxParallelism = 1
+      override def compute(split: Partition, context: TaskContext): Iterator[String] = {
+
+        val validUrl = for {
+          mimeType <- client.files.retrieveByExternalId(fileExternalId).map(_.mimeType)
+          downloadLink <- client.files
+            .downloadLink(FileDownloadExternalId(fileExternalId))
+            .map(_.downloadUrl)
+          _ <- IO.pure(
```

Comment:
```suggestion
          _ <- IO.raiseWhen(
```
looks like there's a dedicated helper for that

Reply:
cool